packages/access/migrations/sqls/20181210164010-improvements-up.sql

```diff
@@ -0,0 +1,62 @@
+-- Function Update: Handle camel-cased table names e.g. "accessGrants".
+--
+-- First argument _tbl accepts of type regclass and requires a table class
+-- either passed with single quotes ('payments'), or double quotes if table name
+-- is camel cased ("accessGrants"). SELECT query in functon later wrapped
+-- _tbl again in double quotes which led to an error.
+--
+CREATE OR REPLACE FUNCTION make_hrid(IN _tbl regclass, IN _column text, IN digits bigint) RETURNS text AS $$
+DECLARE
+chars char[];
+new_hrid text;
+done bool;
+BEGIN
+  chars := ARRAY['1','2','3','4','5','6','7','8','9','A','B','C','D','E','F','G','H','K','L','M','R','S','T','U','W','X','Y','Z'];
+  done := false;
+  <<doneloop>>
+  WHILE NOT done LOOP
+    new_hrid := '';
+    <<hridloop>>
+    WHILE char_length(new_hrid) < digits LOOP
+      new_hrid := new_hrid || chars[floor(random()*(array_length(chars, 1)-1+1))+1];
+    END LOOP hridloop;
+    EXECUTE format('SELECT (NOT EXISTS (SELECT 1 FROM %s WHERE "%s" = %L))::bool', _tbl, _column, new_hrid) INTO done;
+  END LOOP doneloop;
+  RETURN new_hrid;
+END;
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+ALTER TABLE "accessGrants"
+  ALTER COLUMN "beginAt" DROP NOT NULL,
+  ALTER COLUMN "endAt" DROP NOT NULL,
+  ADD COLUMN "beginBefore" timestamp with time zone,
+  ADD COLUMN "voucherCode" text UNIQUE,
+  ADD COLUMN "message" text
+;
+
+-- Set voucherCode to unclaimed but still valid grants.
+UPDATE "accessGrants"
+SET "voucherCode" = make_hrid('"accessGrants"', 'voucherCode', 5)
+WHERE "recipientUserId" IS NULL AND "invalidatedAt" IS NULL ;
+
+ALTER TABLE "accessGrants"
+  ALTER COLUMN "voucherCode"
+    SET DEFAULT make_hrid('"accessGrants"', 'voucherCode', 5)
+;
+
+ALTER TABLE "accessCampaigns"
+  ADD COLUMN "validInterval" interval NOT NULL DEFAULT '30 days'::interval,
```

Comment:
Describing time is a shitty thing. We may just have to agree. Here's what I believe to be concise enough for us, taking existing data structure into account:

*An interval describes a duration, length or amount of time* e.g. "30 days 12 hours".

It is relative, and neither a point in time ("2018-12-12 12:00", a timestamp) nor has it a specific beginning and ending ("2018-12-12 15:00 to 2019-01-12 15:00", a period).

An interval can be construed as a measure between date A and date B. And it may help determine either dates in case date A or date B is missing e.g. `date A + interval = date B`.

Postgres thinks along the same lines e.g. `'30 days 12 hours'::interval` which lets me to believe naming attributes with "interval" is reasonable. And it makes good use of it: Postres' `age(start, end)` returns an interval. Pogi converts an Postgres interval [field into a `PostgresInterval` object](https://www.npmjs.com/package/postgres-interval).

A *period* consists of a beginning and an end date e.g. "2018-12-12 15:00 to 2019-01-12 15:00".

`membershipPeriods` in backends contains records of start and end date. An interval can be used to create a period. This is again where I find Postgres' `interval` utterly handy.

And now to the "time is shitty" part, in the spirit of full disclosure: Recently introduce `membershipTypes.interval`, `membershipTypes.minPeriods` etc. contradicts aforementioned ideas already somewhat.

And then there is Moment.JS: [It's pendent to "intervals" is "durations"](https://momentjs.com/docs/#/durations/), very much along the lines of [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Durations). Said standard [also mentions "time intervals"](https://en.wikipedia.org/wiki/ISO_8601#Time_intervals) describing them as a compound of periods and durations, including "Repeating intervals".

Pretty sure you have some thought's on this, too?

Reply:
Meanwhile to increase readability and understanding, settled to rename fields:

* `validInterval` becomes `grantClaimableInterval`, generating an implicit period in which a grant can be claimed. Implicit because only end date of period is stored in `grant.beginBefore`, start date being `grant.createdAt`.
* `periodInterval` becomes `grantPeriodInterval`, generating an explicit period with `grant.beginAt` and `grant.endAt`.

Hope this is well in your spirit.